File: esj-rules/src/main/java/de/bsnsoft/esj/rules/CodeLists.java
package de.bsnsoft.esj.rules;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;

/**
 * The code list snapshots one pack was compiled with, keyed by the identifier a rule names
 * a list by.
 *
 * <p>Which snapshot that is, is decided by the pack manifest and by nothing else: a rule
 * writes {@code {"inList": [ {"value": "/BT-5"}, "iso-4217" ]}} and the manifest says that
 * {@code iso-4217} in this pack means the file of one particular day. Two runs of the same
 * pack version over the same document therefore give the same verdict however far apart
 * they are, which is the only way a validation report keeps its meaning in an archive.
 *
 * <p>A pack that names a list it has no snapshot for does not compile. That is deliberate
 * and it is the reason this class refuses rather than returns an empty list: a membership
 * test against a list nobody loaded would pass every code, and a rule that silently passes
 * everything is worse than a pack that will not start.
 */
public final class CodeLists {

    /** Where a bundled pack keeps its snapshots, below the resource directory of a pack. */
    private static final String CODELISTS = "codelists/";

    private final Map<String, CodeList> byId;

    private CodeLists(Map<String, CodeList> byId) {
        this.byId = byId;
    }

    /**
     * Collects snapshots that are already in hand.
     *
     * @param lists the snapshots; two of the same identifier are a mistake and are refused
     * @return the collection
     * @throws RulePackException    if two snapshots carry the same identifier
     * @throws NullPointerException if {@code lists} or an element is {@code null}
     */
    public static CodeLists of(Collection<CodeList> lists) {
        Objects.requireNonNull(lists, "lists");
        Map<String, CodeList> byId = new LinkedHashMap<>();
        for (CodeList list : lists) {
            if (byId.put(list.listId(), list) != null) {
                throw new RulePackException("two snapshots carry the identifier " + list.listId());
            }
        }
        return new CodeLists(Collections.unmodifiableMap(byId));
    }

    /**
     * Collects no snapshots at all, for a pack whose rules ask about none.
     *
     * @return the empty collection
     */
    public static CodeLists empty() {
        return new CodeLists(Map.of());
    }

    /**
     * Loads the snapshots a pack manifest names, from the rule directory that travels in
     * this module's jar.
     *
     * @param pack the pack whose manifest names the snapshots
     * @return the snapshots, keyed by list identifier
     * @throws RulePackException    if a snapshot the manifest names is not in the jar or is
     *                              not a snapshot
     * @throws NullPointerException if {@code pack} is {@code null}
     */
    public static CodeLists bundled(RulePack pack) {
        Objects.requireNonNull(pack, "pack");
        List<CodeList> lists = new ArrayList<>();
        for (Map.Entry<String, String> named : pack.codeLists().entrySet()) {
            String resource = RulePacks.resourceDirectory(pack.id(), pack.version())
                    + CODELISTS + named.getKey() + "/" + named.getValue() + ".json";
            try (InputStream in = CodeLists.class.getResourceAsStream(resource)) {
                if (in == null) {
                    throw new RulePackException("the pack " + pack.id() + " " + pack.version()
                            + " names the snapshot " + named.getKey() + " of " + named.getValue()
                            + ", which is not in this build");
                }
                lists.add(read(in, named.getKey() + " of " + named.getValue()));
            } catch (IOException e) {
                throw new RulePackException("the snapshot " + resource + " could not be read: "
                        + e.getMessage(), e);
            }
        }
        return of(lists);
    }

    /**
     * Reads one snapshot.
     *
     * @param in   the bytes of the snapshot file, UTF-8; the caller owns the stream
     * @param what what is being read, for the message of a failure
     * @return the snapshot
     * @throws RulePackException if the bytes are not a snapshot
     */
    public static CodeList read(InputStream in, String what) {
        Json json = Json.read(in, "the code list snapshot " + what);
        Map<String, Json> members = json.asObject("the code list snapshot " + what);
        Set<String> known = Set.of("listId", "name", "publisher", "source", "retrieved", "terms", "entries");
        for (String member : members.keySet()) {
            if (!known.contains(member)) {
                throw new RulePackException("the code list snapshot " + what
                        + " carries the member " + member + ", which this format does not define");
            }
        }
        Map<String, String> entries = new LinkedHashMap<>();
        for (Json entry : required(members, "entries", what).asArray("entries of " + what)) {
            Map<String, Json> fields = entry.asObject("an entry of " + what);
            String value = required(fields, "value", what).asString("the value of an entry of " + what);
            String name = fields.containsKey("name")
                    ? fields.get("name").asString("the name of an entry of " + what)
                    : "";
            if (entries.put(value, name) != null) {
                throw new RulePackException("the code list snapshot " + what
                        + " carries the code " + value + " twice");
            }
        }
        return new CodeList(
                required(members, "listId", what).asString("listId of " + what),
                required(members, "name", what).asString("name of " + what),
                required(members, "publisher", what).asString("publisher of " + what),
                required(members, "source", what).asString("source of " + what),
                required(members, "retrieved", what).asString("retrieved of " + what),
                entries);
    }

    private static Json required(Map<String, Json> members, String name, String what) {
        Json value = members.get(name);
        if (value == null) {
            throw new RulePackException("the code list snapshot " + what + " has no " + name);
        }
        return value;
    }

    /**
     * Returns the snapshot a rule names, if this collection has it.
     *
     * @param listId the identifier a rule names the list by
     * @return the snapshot, or an empty optional
     */
    public Optional<CodeList> list(String listId) {
        return Optional.ofNullable(byId.get(listId));
    }

    /**
     * Returns the snapshot a rule names, or refuses.
     *
     * @param listId the identifier a rule names the list by
     * @return the snapshot
     * @throws RulePackException if this collection has no snapshot of that identifier
     */
    public CodeList require(String listId) {
        CodeList list = byId.get(listId);
        if (list == null) {
            throw new RulePackException("no snapshot of the code list " + listId
                    + " was loaded; a membership test against a list nobody loaded would pass"
                    + " every code, so the pack does not compile");
        }
        return list;
    }

    /**
     * Returns the identifiers of the snapshots in this collection.
     *
     * @return the identifiers, in the order the snapshots were collected
     */
    public Set<String> listIds() {
        return byId.keySet();
    }
}

File: bindings/csharp/En16931.SemanticJson/Rules/CodeLists.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace En16931.SemanticJson.Rules;

/// <summary>
/// A dated snapshot of a code list a rule decides membership against.
/// </summary>
/// <remarks>
/// A rule asks <c>inList</c> about a list identifier and the pack manifest says which day's
/// snapshot of that list this pack decides against, so two runs of the same pack version over
/// the same document give the same verdict however far apart they are. That is the only way a
/// validation report keeps its meaning in an archive.
/// </remarks>
public sealed class CodeList
{
    private readonly Dictionary<string, string> _entries;

    internal CodeList(
        string listId, string name, string publisher, string source, string retrieved,
        Dictionary<string, string> entries)
    {
        ListId = listId;
        Name = name;
        Publisher = publisher;
        Source = source;
        Retrieved = retrieved;
        _entries = entries;
    }

    /// <summary>Returns the identifier a rule names the list by.</summary>
    public string ListId { get; }

    /// <summary>Returns the name of the list.</summary>
    public string Name { get; }

    /// <summary>Returns who publishes it.</summary>
    public string Publisher { get; }

    /// <summary>Returns where the snapshot was taken from.</summary>
    public string Source { get; }

    /// <summary>Returns the day the snapshot was taken.</summary>
    public string Retrieved { get; }

    /// <summary>Returns how many codes the snapshot holds.</summary>
    public int Size => _entries.Count;

    /// <summary>Tells whether a code is on the snapshot.</summary>
    /// <param name="code">the code</param>
    /// <returns>whether the list carries it</returns>
    public bool Contains(string code)
    {
        ArgumentNullException.ThrowIfNull(code);
        return _entries.ContainsKey(code);
    }

    /// <summary>Returns what the publisher calls a code, or <c>null</c>.</summary>
    /// <param name="code">the code</param>
    /// <returns>the name, or <c>null</c> where the list does not carry the code</returns>
    public string? Describe(string code) => _entries.TryGetValue(code, out string? name) ? name : null;

    /// <inheritdoc />
    public override string ToString() =>
        ListId + " (" + Publisher + ", " + Retrieved + ", " + _entries.Count + " codes)";
}

/// <summary>The snapshots one pack decides membership against.</summary>
public sealed class CodeLists
{
    private readonly Dictionary<string, CodeList> _byId;

    private CodeLists(Dictionary<string, CodeList> byId)
    {
        _byId = byId;
    }

    /// <summary>Returns a set of snapshots.</summary>
    /// <param name="lists">the snapshots</param>
    /// <returns>the set</returns>
    public static CodeLists Of(IEnumerable<CodeList> lists)
    {
        ArgumentNullException.ThrowIfNull(lists);
        Dictionary<string, CodeList> byId = new(StringComparer.Ordinal);
        foreach (CodeList list in lists)
        {
            if (!byId.TryAdd(list.ListId, list))
            {
                throw new RulePackException("two snapshots carry the identifier " + list.ListId);
            }
        }

        return new CodeLists(byId);
    }

    /// <summary>Returns no snapshots at all.</summary>
    /// <returns>the empty set</returns>
    public static CodeLists Empty() => new(new Dictionary<string, CodeList>(StringComparer.Ordinal));

    /// <summary>Returns the snapshots a bundled pack names, as this build carries them.</summary>
    /// <param name="pack">the pack</param>
    /// <returns>the snapshots</returns>
    /// <exception cref="RulePackException">if a snapshot the pack names is not in this build</exception>
    public static CodeLists Bundled(RulePack pack)
    {
        ArgumentNullException.ThrowIfNull(pack);
        List<CodeList> lists = new();
        foreach (KeyValuePair<string, string> named in pack.CodeLists)
        {
            string resource = RulePacks.ResourceDirectory(pack.Id, pack.Version)
                + "codelists/" + named.Key + "/" + named.Value + ".json";
            using Stream stream = RulePacks.Open(resource);
            lists.Add(Read(stream, named.Key + " of " + named.Value));
        }

        return Of(lists);
    }

    /// <summary>Reads one snapshot.</summary>
    /// <param name="input">the snapshot file</param>
    /// <param name="what">what is being read, for the message of a failure</param>
    /// <returns>the snapshot</returns>
    public static CodeList Read(Stream input, string what)
    {
        ArgumentNullException.ThrowIfNull(input);
        using JsonDocument json = JsonDocument.Parse(input);
        JsonElement root = json.RootElement;
        Dictionary<string, string> entries = new(StringComparer.Ordinal);
        foreach (JsonElement entry in root.GetProperty("entries").EnumerateArray())
        {
            string value = entry.GetProperty("value").GetString()!;
            string name = entry.TryGetProperty("name", out JsonElement written)
                ? written.GetString() ?? string.Empty
                : string.Empty;
            if (!entries.TryAdd(value, name))
            {
                throw new RulePackException(
                    "the code list snapshot " + what + " carries the code " + value + " twice");
            }
        }

        return new CodeList(
            root.GetProperty("listId").GetString()!,
            root.GetProperty("name").GetString()!,
            root.GetProperty("publisher").GetString()!,
            root.GetProperty("source").GetString()!,
            root.GetProperty("retrieved").GetString()!,
            entries);
    }

    /// <summary>Returns the identifiers of the snapshots this set holds.</summary>
    public IReadOnlyCollection<string> ListIds => _byId.Keys.ToList();

    /// <summary>Returns a snapshot, or <c>null</c> where this set holds none.</summary>
    /// <param name="listId">the identifier of the list</param>
    /// <returns>the snapshot, or <c>null</c></returns>
    public CodeList? List(string listId)
    {
        ArgumentNullException.ThrowIfNull(listId);
        return _byId.TryGetValue(listId, out CodeList? list) ? list : null;
    }

    /// <summary>
    /// Returns a snapshot, or refuses: a membership test against a list nobody loaded would
    /// pass every code, and a rule that silently passes everything is worse than a pack that
    /// will not start.
    /// </summary>
    /// <param name="listId">the identifier of the list</param>
    /// <returns>the snapshot</returns>
    /// <exception cref="RulePackException">if this set holds no such snapshot</exception>
    public CodeList Require(string listId) => List(listId)
        ?? throw new RulePackException("no snapshot of the code list " + listId + " was loaded");
}
